### ISSUE TITLE
Method for easily creating associations

### ISSUE THREAD

**Issue by paulcsmith** [luckyframework/migrator#22]
This will be called `belongs_to` to mimic what it looks like in models in LuckyRecord:

```crystal
create comments do
  # Will create an Int64 column named user_id, with a foreign key, and an index
  # Will be null false unless you do `belongs_to User?`
  belongs_to User 
end
```

**Comment by mikeeus** [comment]
Are you thinking of something like this?

```crystal
it "can create associations" do
    built = LuckyMigrator::CreateTableStatement.new(:comments).build do
      belongs_to : User, cascade: true
      belongs_to : Post, restrict: true
      belongs_to : Category?
      add body : String
    end

    built.table_statement.should eq <<-SQL
    CREATE TABLE comments (
      id serial PRIMARY KEY,
      created_at timestamp NOT NULL,
      updated_at timestamp NOT NULL,
      user_id bigint NOT NULL REFERENCES users ON DELETE CASCADE,
      post_id bigint NOT NULL REFERENCES posts ON DELETE RESTRICT,
      category_id bigint REFERENCES category,      
      body text NOT NULL)
    SQL

    built.index_statements.first.should eq "  CREATE INDEX comments_user_id_index ON comments USING btree (user_id)"
    built.index_statements[1].should eq "  CREATE INDEX comments_post_id_index ON comments USING btree (post_id)"
    built.index_statements.last.should eq "  CREATE INDEX comments_category_id_index ON comments USING btree (category_id)"    
  end
```
This is dependent on #17 

I'm guessing the model types would be required automatically from LuckyRecord, right?

**Comment by paulcsmith** [comment]
@mikeeus Yep that is exactly right!  I think for the first pass it might be good to leave off the `ON DELETE` stuff to keep the PR simple. But it could be added in a follow up PR

Some minor syntax changes though:

```crystal
belongs_to User # Doesn't need the colon
```

Also the `belongs_to` would be a macro. So the class that is passed in does not need to exist. It's mostly just there to get the foreign key name and determine if it should be optional:

```crystal
# Rough version of what it would do
macro belongs_to(model_class)
  {% foreign_key_name = (model_class.stringify.gsub(/\?/, "").underscore + "_id") %}
  {% optional = model_class.stringify.ends_with?("?") %}
  p {{ foreign_key_name }}
  p {{ optional }}
end

belongs_to User
```

https://play.crystal-lang.org/#/r/33ml

Does that help answer your question?

**Comment by mikeeus** [comment]
Okay cool, I see how that works. I'll give it a shot

**Comment by paulcsmith** [comment]
Awesome! I'm so grateful for your help on all this :)

**Comment by mikeeus** [comment]
I need to pluralize the model names to reference the table. Is there a pluralize method in the std lib or should I add a basic implementation for this?

**Comment by paulcsmith** [comment]
@mikeeus I think you should be able to get the table name because `@table_name` is available in both the CreateTableStatement and AlterTableStatement. So you might do something like this:

```crystal
macro belongs_to(model_name)
  # Get the foreign key name and whether it is optional then...
  add_belongs_to({{ foreign_key_name }}, optional: {{ optional }}
end

def add_belongs_to(foreign_key_name : String, optional : Bool)
  # Use @table_name here to get the table name to reference
end
```

**Comment by mikeeus** [comment]
I meant the table being referenced, so that belongs_to User can generate the " REFERENCES users" string.
```
create :comments do
    belongs_to User
end

CREATE TABLE comments (
        ...
        user_id bigint NOT NULL REFERENCES users,)
      SQL
```

**Comment by paulcsmith** [comment]
@mikeeus Oh I see what you mean. I think for now let's just add an `s` to the underscored class name. @actsasflinn is working on some pluralization stuff that we can use later so that it is a bit more robust.

Let's also add a `references` option so people can provide a custom table name if they need to: `belongs_to Author, references: :users`

**Comment by mikeeus** [comment]
I was looking at [https://github.com/rails/rails/blob/48b37f127fb44b787ae6c0710982d7949b14454f/activesupport/lib/active_support/inflections.rb](active records inflector), is he looking at something like that?

And that references option is a good idea, I'll add that

**Comment by paulcsmith** [comment]
@mikeeus Yep! He's basing it heavily off of that and will be making a PR to https://github.com/luckyframework/support soon.

**Comment by mikeeus** [comment]
Great, I'll keep it simple until that PR goes up

**Comment by mikeeus** [comment]
kk I made a PR for this. I'm done for today so let me know what needs to be done and I'll take care of it tomorrow

**Comment by actsasflinn** [comment]
Awesome, this kind of stuff is what I had in mind in tackling the Inflector. I have one more feature from Rails to port which is parameterize from transliterate.rb before I PR. If you are looking for pluralize right now it’s in good shape. I have it in the inflector branch. 

https://github.com/actsasflinn/support/tree/inflector

**Comment by mikeeus** [comment]
@actsasflinn Nice! That's looking really good. I'm sure I'll find lots of uses for it

**Comment by paulcsmith** [comment]
Closed by #25
